File: rcm-kit/src/menus/autorun.ts
import { ADD_TO_AUTORUN, REMOVE_FROM_AUTORUN } from "../consts"
import { t } from "../i18n"
import type { MenuItem, InvokeProps } from "../types"

/**
 * Win11 "Add to Startup" / "Remove from Startup" —
 * adds or removes a .exe file from the Windows startup (autorun) list
 * via the registry Run keys.
 *
 * The Rust backend provides `inAutorun` (entry name list) so the
 * frontend can check whether the selected .exe is already registered.
 *
 * Only matches single .exe file selections (not .lnk, not folders).
 */

/** True when the selected file is an .exe (single selection). */
function isExe(props: InvokeProps): boolean {
  if (props.files.length !== 1) return false
  return props.files[0].path.toLowerCase().endsWith(".exe")
}

/** Extract file stem from a Windows path (name without extension). */
function fileStem(path: string): string {
  const name = path.split(/[\\/]/).pop() ?? path
  const dot = name.lastIndexOf(".")
  return dot > 0 ? name.slice(0, dot) : name
}

/** Extract the bare .exe path from a registry command string.
 *  Handles quoted paths, trailing \\0, and extra arguments. */
function exePath(command: string): string {
  // Strip trailing null bytes (registry REG_SZ sometimes includes them)
  let cmd = command.replace(/\0/g, "")
  // If the command starts with a quote, extract the quoted part
  if (cmd.startsWith('"')) {
    const end = cmd.indexOf('"', 1)
    if (end > 0) cmd = cmd.slice(1, end)
  } else {
    // Otherwise take the first space-delimited token
    cmd = cmd.split(/\s+/)[0]
  }
  return cmd
}

function isInAutorun(props: InvokeProps): boolean {
  const target = props.files[0].path.toLowerCase()
  return props.autorun.some((e) => exePath(e.command).toLowerCase() === target)
}

/**
 * "Add to Startup" — shown when the .exe is NOT already in autorun.
 * @param label Custom label, defaults to system i18n text.
 */
export function addToAutorun(label = t("add.to.startup")): MenuItem {
  return {
    key: "add-to-autorun",
    label,
    icon: "🚀",
    match: (props) => isExe(props) && !isInAutorun(props),
    action: (props) => ({
      cmd: ADD_TO_AUTORUN,
      args: [fileStem(props.files[0].path), props.files[0].path],
      window: "Hidden",
    }),
  }
}

/**
 * "Remove from Startup" — shown when the .exe IS already in autorun.
 * @param label Custom label, defaults to system i18n text.
 */
export function removeFromAutorun(label = t("remove.from.startup")): MenuItem {
  return {
    key: "remove-from-autorun",
    label,
    icon: "🚀",
    match: (props) => isExe(props) && isInAutorun(props),
    action: (props) => ({
      cmd: REMOVE_FROM_AUTORUN,
      args: [props.files[0].path],
      window: "Hidden",
    }),
  }
}

File: rcm-core/src/cmds/autorun.rs
//! `@add-to-autorun` / `@remove-from-autorun` — Add or remove an .exe
//! file from the Windows startup (autorun) list.
//!
//! Uses the `autorun` crate to manipulate the registry Run keys
//! (`HKCU\...\Run` and `HKLM\...\Run`). Only `User` scope is
//! used for add/remove (no elevation required).
//!
//! The frontend sends the file stem as `name` (arg[0]) and the
//! full path as `command` (arg[1]) for add; the full path alone for remove
//! (backend resolves name by matching the command).

use autorun::Entry;

use super::SystemCmdResult;
use crate::types::CommandPayload;

/// Run `@add-to-autorun` — add a program to Windows startup.
///
/// Expects `args[0]` = entry name (file stem), `args[1]` = full path.
pub fn run_add(cmd: &CommandPayload) -> SystemCmdResult {
    let (name, command) = match (cmd.args.first(), cmd.args.get(1)) {
        (Some(n), Some(c)) if !n.is_empty() && !c.is_empty() => (n.as_str(), c.as_str()),
        _ => {
            return SystemCmdResult {
                success: false,
                message: "Requires name and command arguments".into(),
            };
        }
    };

    crate::log::info(
        "Rust::add_to_autorun",
        &format!("adding '{name}' → '{command}' to startup"),
    );

    match autorun::add(name, command, autorun::Scope::User) {
        Ok(()) => {
            crate::log::info("Rust::add_to_autorun", "add OK");
            SystemCmdResult {
                success: true,
                message: format!("Added to startup: {name}"),
            }
        }
        Err(e) => {
            let msg = e.to_string();
            crate::log::error("Rust::add_to_autorun", &msg);
            SystemCmdResult {
                success: false,
                message: msg,
            }
        }
    }
}

/// Run `@remove-from-autorun` — remove a program from Windows startup.
///
/// Expects `args[0]` = full path of the .exe (matched against stored commands).
pub fn run_remove(cmd: &CommandPayload) -> SystemCmdResult {
    let path = match cmd.args.first() {
        Some(p) if !p.is_empty() => p.as_str(),
        _ => {
            return SystemCmdResult {
                success: false,
                message: "No path specified".into(),
            };
        }
    };

    // Find the entry by matching the command path.
    let entry = match find_entry_name_by_command(path) {
        Some(e) => e,
        None => {
            return SystemCmdResult {
                success: false,
                message: format!("Not found in startup: {path}"),
            };
        }
    };

    crate::log::info(
        "Rust::remove_from_autorun",
        &format!("removing '{}' ({path}) from startup", entry.name),
    );

    match entry.remove() {
        Ok(()) => {
            crate::log::info("Rust::remove_from_autorun", "remove OK");
            SystemCmdResult {
                success: true,
                message: format!("Removed from startup: {path}"),
            }
        }
        Err(e) => {
            let msg = e.to_string();
            crate::log::error("Rust::remove_from_autorun", &msg);
            SystemCmdResult {
                success: false,
                message: msg,
            }
        }
    }
}

/// List all autorun entries from both HKCU and HKLM scopes.
///
/// The frontend uses this list to decide whether a given .exe is
/// already in the startup list (matched by `command` = full path).
pub fn list_autorun_entries() -> Vec<Entry> {
    autorun::list().unwrap_or_else(|e| {
        crate::log::error("Rust::list_autorun", &e.to_string());
        Vec::new()
    })
}

/// Extract the bare .exe path from a registry command string.
/// Handles quoted paths, trailing NUL bytes, and extra arguments.
fn exe_path(command: &str) -> &str {
    let cmd = command.trim_end_matches('\0');
    if let Some(stripped) = cmd.strip_prefix('"') {
        // Quoted path: take everything until the closing quote
        stripped
            .find('"')
            .map(|i| &stripped[..i])
            .unwrap_or(stripped)
    } else {
        // Unquoted: take the first space-delimited token
        cmd.split_whitespace().next().unwrap_or(cmd)
    }
}

/// Find the autorun entry name that contains the given command path.
fn find_entry_name_by_command(target: &str) -> Option<Entry> {
    autorun::list()
        .ok()?
        .into_iter()
        .find(|e| exe_path(&e.command).eq_ignore_ascii_case(target))
}
